### ISSUE TITLE
Initial feedback

### ISSUE THREAD

**Issue by sepiariver** [ryancrideout/Ryan_Professional_Development#2]
Nice work organizing this! You’re not done yet I think but great start. Some thoughts, in no particular order 

- MapTiles is a good idea if you want each tile to have functionality. Maybe they render in a different colour based on the type of MapTile (subclass) that they are—maybe the terrain they represent. Strictly speaking it is not necessary to achieve the requirements of the exercise just FYI but cool nonetheless 
- I noticed some comments about whether the map controls movement or the characters. I’m no game engineer so take this with a grain of salt…500 grains of salt. It seems to me like the characters must know how they move. If you want to introduce the idea of terrain and the terrain affects movement then you must share that information with the character OR have something “above” it all that orchestrates all character movements across all terrain. I suspect this last approach is how “they” (those who know what they’re doing) do it. ie a “game engine” orchestrates All The Things. Each individual Thing like a character or terrain tile knows nothing about other Things. Have a think on that, if you please :)

**Comment by ryancrideout** [comment]
Hi Hi YJ,

Thanks for leaving your feedback. 

1) The MapTiles just made too much sense for me to not include. I've played a number of strategy games back in my day (well, I still play them) and thinking of maps composed of map tiles just was intuitive for me. Not only could I include colors, but I do different things like put fortresses on tiles, or make them impassable, etc.

2) I puzzled over character movement for a long time, and eventually I landed on the characters knowing _how_ they move, but the game state actually being the one that does the movement. Of course this isn't implemented yet but in my head this is what I was thinking of.

At first I thought of characters being in charge of their movement, but if I wanted to do that with this current set up then each character would need a "Map" attribute or something so they could know what map they're on so they could move themselves. I don't _hate_ that idea, but I feel like that's making the Character class be responsible for something it shouldn't. Having the game keep track of this though seems to make more sense to me.

It's like... in Super Mario 64 (I really hope you know who Mario is), you can have Mario jump, double jump, triple jump, long jump, etc. He has all of these wonderful movement options. BUT, walls and the like can stop him cold (as Mario can't jump through walls... usually). But Mario himself isn't responsible for knowing if there are walls in his way or not - Mario just needs to know how to jump. It's the game that figures out "Okay so Mario wants to jump, is there anything in his way?" and goes from there.

That's my head space at least.

**Comment by sepiariver** [comment]
Yep you’re in the right headspace :)

**Comment by sepiariver** [comment]
Looks like with 2b2182cf152c91a048fee3c80f7ddd3a2861a594 you pretty much stuck to your thinking above, which I think is good! Game operates, it's fun! 

Few thoughts:

- `Character.set_position`'s signature lacks type hints but your child classes have them. Keeps things a bit more tidy to stay consistent
- `Character.movement_action` is a bigger problem because it doesn't have the `direction` parameter and in the child classes where you've added that to the method signatures it is required (without a default value). This violates the the Liskov Substitution Principle
- In Python 3.10 the [match statement](https://docs.python.org/3/whatsnew/3.10.html#pep-634-structural-pattern-matching) was introduced, which functions like a switch statement in other languages. It can make your code a bit easier to understand and cleaner, in cases where you have `if ... elif ... elif ... else` and you're checking the same variable in all cases. Look at `Game.create_character` for this.
- You have comments in multiple places about checking inputs. Good call, I encountered an exception when I tried to move an undefined character :)

This is pretty good and I think it does demonstrate you have some understanding of object-oriented programming and the SOLID principles. We're not going to harp on the subject. I think the next phase will be unittest-ing
